control_protocol.md

```diff
@@ -0,0 +1,478 @@
+API or the Remote Playback API.  The message format should allow each party to
+deserialize each message as it arrives over the transport and dispatch it to
+appropriate code in the implementation of each API.
+
+## Message Flavors
+
+Messages come in four flavors: Commands, Requests, Responses, and Events.
+
+- Commands are unidirectional messages sent from one party to a single
+  recipient.  No response is expected from the recipient.
+- Requests are sent from one party to a single recipient.  That recipient must
+  reply with a Response back to the initial sender that contains a reference to
+  the sequence ID of the Request.
+- Events are sent from a single party to one or more recipients.  No
+  responses are expected from any recipient.
+  
+### Message Structure
+
+Messages are structured as a 36 byte message header, followed by a variable
+length message body.
+
+```
+Byte offset
+  0           +-------------------+
+              +    PROTOCOL_ID    +
+  4           +-------------------+
+              +      FLAGS        +
+  8           +-------------------+
+              +  MESSAGE_LENGTH   +
+  16          +-------------------+
+              +   MESSAGE_TYPE    +
+  24          +-------------------+
+              +   SEQUENCE_ID     +
+  32          +-------------------+
+              +    REQUEST_ID     +
+  36          +-------------------+
+              +   MESSAGE BODY    +
+              .                   .
+              .                   .
+              .                   .
+              +-------------------+
+```
+
+
+The overall length of a message is constrained only by the `MESSAGE_LENGTH`
+field, or 2^64 - 1 bytes.  Practically speaking, messages are limited by the
+memory capacity of each party, the underlying tansport, and the necessity of
+transmitting messages in a reasonable amount of time.
+
+All integers are to be represented in [network byte order](https://tools.ietf.org/html/rfc1700).
+
+
+- `PROTOCOL_ID`: A 32-bit value explained below.
+- `FLAGS`: A 32-bit value explained below.
+- `MESSAGE_LENGTH`: A 64-bit unsigned integer that contains the number of bytes
+  in the entire message (including all header fields).
+- `MESSAGE_TYPE`: A 64-bit value that identifies the message content.
+- `SEQUENCE_ID`: A 64-bit, positive unsigned integer that is used to uniquely
+  identify messages originating from one party, and to ensure that messages are
+  handled in the proper order by the recipient.
+- `REQUEST_ID`: An optional 64-bit unsigned integer.  If the message flavor is a
+  Response, it contains the `SEQUENCE_ID` of the Request that the response is
+  replying to.  For other flavors, it is not present.
+
+The content of the `MESSAGE_BODY` is not constrained by the generic message
+structure, and must be interpreted according to the `MESSAGE_TYPE`.
+
+**TODO:** Investigate variable length integers for MESSAGE_LENGTH, SEQUENCE_ID
```

Comment:
A starting point for the investigation: https://en.wikipedia.org/wiki/Variable-length_quantity. Includes links to particular encodings, e.g, Protocol Buffers. Should we raise as a separate issue for discussion?

Reply:
Good idea. Filed https://github.com/webscreens/openscreenprotocol/issues/45 and will link from this TODO.